File: client/src/wasm/force-calculations.ts
// Type definition for Emscripten module
interface EmscriptenModule {
    _malloc: (size: number) => number;
    _free: (ptr: number) => void;
    HEAPU8: Uint8Array;
    HEAP32: Int32Array;
    NodeVector: {
        new(): NodeVectorType;
    };
    LinkVector: {
        new(): LinkVectorType;
    };
    FloatVector: {
        new(): FloatVectorType;
    };
    ForceSimulation: {
        new(maxDistance: number, velocityDecay: number): ForceSimulation;
    };
}

// Types pour les vecteurs C++
interface NodeVectorType {
    push_back(node: Node): void;
    get(index: number): Node;
    size(): number;
    delete(): void;
}

interface LinkVectorType {
    push_back(link: Link): void;
    get(index: number): Link;
    size(): number;
    delete(): void;
}

interface FloatVectorType {
    push_back(value: number): void;
    get(index: number): number;
    size(): number;
    delete(): void;
}

interface Vec3 {
    x: number;
    y: number;
    z: number;
}

interface Node {
    position: Vec3;
    velocity: Vec3;
    charge: number;
}

interface Link {
    source: number;
    target: number;
}

interface ForceSimulation {
    setNodes(nodes: NodeVectorType): void;
    setLinks(links: LinkVectorType): void;
    setDistances(distances: FloatVectorType): void;
    setStrengths(strengths: FloatVectorType): void;
    getNodes(): NodeVectorType;
    step(): void;
}

interface ForceCalculationsModule extends EmscriptenModule {
    ForceSimulation: {
        new(maxDistance: number, velocityDecay: number): ForceSimulation;
    };
}

declare global {
    interface Window {
        createModule?: () => Promise<ForceCalculationsModule>;
    }
}

let moduleInstance: ForceCalculationsModule | null = null;
let simulation: ForceSimulation | null = null;

function loadWasmModule(): Promise<void> {
    return new Promise((resolve, reject) => {
        const script = document.createElement('script');
        script.src = '/wasm/force-calculations.js';
        script.onload = () => resolve();
        script.onerror = (error) => reject(new Error(`Failed to load WASM script: ${error}`));
        document.head.appendChild(script);
    });
}

export async function initForceCalculations(): Promise<ForceCalculationsModule> {
    if (moduleInstance) {
        return moduleInstance;
    }

    try {
        await loadWasmModule();

        if (typeof window.createModule !== 'function') {
            throw new Error('createModule is not available. Make sure the WASM module is loaded correctly.');
        }

        moduleInstance = await window.createModule();
        simulation = new moduleInstance.ForceSimulation(100, 0.1);
        console.log('WebAssembly module initialized successfully');
        return moduleInstance;
    } catch (error) {
        console.error('Failed to initialize WebAssembly module:', error);
        throw error;
    }
}

export function calculateForces(
    nodes: Node[],
    links: [number, number][],
    distances: number[],
    strengths: number[]
): void {
    if (!simulation || !moduleInstance) {
        throw new Error('WebAssembly module not initialized. Call initForceCalculations first.');
    }

    try {
        // Create C++ vectors
        const nodeVector = new moduleInstance.NodeVector();
        const linkVector = new moduleInstance.LinkVector();
        const distanceVector = new moduleInstance.FloatVector();
        const strengthVector = new moduleInstance.FloatVector();

        // Fill node vector with current state
        nodes.forEach(node => {
            // Ensure velocity is bounded to prevent instability
            const maxVelocity = 10;
            if (node.velocity) {
                node.velocity.x = Math.max(Math.min(node.velocity.x, maxVelocity), -maxVelocity);
                node.velocity.y = Math.max(Math.min(node.velocity.y, maxVelocity), -maxVelocity);
                node.velocity.z = Math.max(Math.min(node.velocity.z, maxVelocity), -maxVelocity);
            }
            nodeVector.push_back(node);
        });

        // Fill link vector with normalized indices
        links.forEach(([source, target]) => {
            // Ensure indices are valid
            if (source >= 0 && source < nodes.length && target >= 0 && target < nodes.length) {
                linkVector.push_back({ source, target });
            }
        });

        // Fill distance vector with normalized values
        distances.forEach(distance => {
            // Ensure distance is positive and reasonable
            const normalizedDistance = Math.max(distance, 1);
            distanceVector.push_back(normalizedDistance);
        });

        // Fill strength vector with normalized values
        strengths.forEach(strength => {
            // Ensure strength is bounded
            const normalizedStrength = Math.max(Math.min(strength, 1), 0);
            strengthVector.push_back(normalizedStrength);
        });

        // Update simulation with vectors
        simulation.setNodes(nodeVector);
        simulation.setLinks(linkVector);
        simulation.setDistances(distanceVector);
        simulation.setStrengths(strengthVector);

        // Run simulation step
        simulation.step();

        // Get updated nodes and carefully update positions
        const updatedNodes = simulation.getNodes();
        for (let i = 0; i < nodes.length; i++) {
            const updatedNode = updatedNodes.get(i);
            if (updatedNode) {
                // Apply position updates with damping
                const damping = 0.8; // Reduce the effect of position changes
                nodes[i].position.x += (updatedNode.position.x - nodes[i].position.x) * damping;
                nodes[i].position.y += (updatedNode.position.y - nodes[i].position.y) * damping;
                nodes[i].position.z += (updatedNode.position.z - nodes[i].position.z) * damping;
                
                // Update velocities with the same damping
                nodes[i].velocity = {
                    x: updatedNode.velocity.x * damping,
                    y: updatedNode.velocity.y * damping,
                    z: updatedNode.velocity.z * damping
                };
            }
        }

        // Clean up vectors
        nodeVector.delete();
        linkVector.delete();
        distanceVector.delete();
        strengthVector.delete();
    } catch (error) {
        console.error('Error in force calculation:', error);
        throw error;
    }
} 

File: client/src/wasm/force-calculations.cpp
#include <emscripten/bind.h>
#include <emscripten/emscripten.h>
#include <vector>
#include <cmath>
#include <memory>

using namespace emscripten;

struct Vec3 {
    float x, y, z;
};

struct Node {
    Vec3 position;
    Vec3 velocity;
    float charge;
};

struct Link {
    int source;
    int target;
};

class ForceSimulation {
private:
    std::vector<Node> nodes;
    std::vector<Link> links;
    std::vector<float> distances;
    std::vector<float> strengths;
    float maxDistance;
    float velocityDecay;

public:
    ForceSimulation(float maxDist = 100.0f, float velDecay = 0.1f)
        : maxDistance(maxDist), velocityDecay(velDecay) {}

    void setNodes(const std::vector<Node>& newNodes) {
        nodes = newNodes;
    }

    void setLinks(const std::vector<Link>& newLinks) {
        links = newLinks;
    }

    void setDistances(const std::vector<float>& newDistances) {
        distances = newDistances;
    }

    void setStrengths(const std::vector<float>& newStrengths) {
        strengths = newStrengths;
    }

    const std::vector<Node>& getNodes() const {
        return nodes;
    }

    void step() {
        // Calcul des forces de répulsion
        for (size_t i = 0; i < nodes.size(); i++) {
            Vec3 force = {0, 0, 0};
            for (size_t j = 0; j < nodes.size(); j++) {
                if (i != j) {
                    float dx = nodes[j].position.x - nodes[i].position.x;
                    float dy = nodes[j].position.y - nodes[i].position.y;
                    float dz = nodes[j].position.z - nodes[i].position.z;
                    
                    float distance = sqrt(dx*dx + dy*dy + dz*dz);
                    if (distance > 0 && distance < maxDistance) {
                        float repulsion = nodes[i].charge * nodes[j].charge / (distance * distance);
                        force.x += dx/distance * repulsion;
                        force.y += dy/distance * repulsion;
                        force.z += dz/distance * repulsion;
                    }
                }
            }
            nodes[i].velocity.x += force.x;
            nodes[i].velocity.y += force.y;
            nodes[i].velocity.z += force.z;
        }

        // Calcul des forces des liens
        for (size_t i = 0; i < links.size(); i++) {
            int source = links[i].source;
            int target = links[i].target;
            float targetDistance = distances[i];
            float strength = strengths[i];
            
            float dx = nodes[target].position.x - nodes[source].position.x;
            float dy = nodes[target].position.y - nodes[source].position.y;
            float dz = nodes[target].position.z - nodes[source].position.z;
            
            float distance = sqrt(dx*dx + dy*dy + dz*dz);
            if (distance > 0) {
                float force = strength * (distance - targetDistance);
                float fx = dx/distance * force;
                float fy = dy/distance * force;
                float fz = dz/distance * force;
                
                nodes[source].velocity.x += fx;
                nodes[source].velocity.y += fy;
                nodes[source].velocity.z += fz;
                
                nodes[target].velocity.x -= fx;
                nodes[target].velocity.y -= fy;
                nodes[target].velocity.z -= fz;
            }
        }

        // Mise à jour des positions
        for (auto& node : nodes) {
            node.position.x += node.velocity.x;
            node.position.y += node.velocity.y;
            node.position.z += node.velocity.z;
            
            node.velocity.x *= velocityDecay;
            node.velocity.y *= velocityDecay;
            node.velocity.z *= velocityDecay;
        }
    }
};

// Bindings pour l'interface avec JavaScript
EMSCRIPTEN_BINDINGS(force_calculations) {
    value_object<Vec3>("Vec3")
        .field("x", &Vec3::x)
        .field("y", &Vec3::y)
        .field("z", &Vec3::z);
        
    value_object<Node>("Node")
        .field("position", &Node::position)
        .field("velocity", &Node::velocity)
        .field("charge", &Node::charge);

    value_object<Link>("Link")
        .field("source", &Link::source)
        .field("target", &Link::target);
        
    register_vector<Node>("NodeVector");
    register_vector<Link>("LinkVector");
    register_vector<float>("FloatVector");

    class_<ForceSimulation>("ForceSimulation")
        .constructor<float, float>()
        .function("setNodes", &ForceSimulation::setNodes)
        .function("setLinks", &ForceSimulation::setLinks)
        .function("setDistances", &ForceSimulation::setDistances)
        .function("setStrengths", &ForceSimulation::setStrengths)
        .function("getNodes", &ForceSimulation::getNodes)
        .function("step", &ForceSimulation::step);
} 